Replace vagrant-based testing with docker-based testing

RaghavaAlajangi opened the pull request:
This PR aims to replace vagrant-based testing with docker-based testing. Docker-based testing is useful because it is faster to execute, easier to manage dependencies, and better integrated with CI/CD.

### To do:
- [x] Make changes in the CICD file
- [x] Add `docker-run-test.sh` and `docker-compose.ci.yml` files
- [ ] Tests pass locally using docker
- [x] Update CHANGELOG
- [x] Passed CICD

RaghavaAlajangi commented:
Hi @paulmueller 
can you review my changes?

paulmueller reviewed (changes requested):
Currently, the CI pipeline passes. However, the tests do not pass. When the tests fail, the pipeline should fail as well.

RaghavaAlajangi commented:
Ohh..I did not see that!

RaghavaAlajangi commented:
Hi @paulmueller 


> Currently, the CI pipeline passes. However, the tests do not pass. When the tests fail, the pipeline should fail as well.

I fixed this issue. However, if you see the failed tests, there is a problem with the license path:
```
license_url = 'file:///srv/app/.local/lib/python3.10/site-packages/ckanext/dcor_schemas/licenses.json'
```
I set this variable while creating the image. But, now the `.local` folder disappeared.

when I run the tests locally, I manually copy this file into the container to see if it works. Then, I get a new error. 

```
botocore.exceptions.EndpointConnectionError: Could not connect to the endpoint URL: "http://127.0.0.1:9000/"
```
and 

```
resource = {'name': 'test.rtdc', 'package_id': '2e6f9733-01af-4fd4-be89-ca655d02c886', 'upload': <FakeFileStorage: 'test.rtdc' (None)>, 'url': 'upload'}

    def get_base_metadata(resource):
        res_dict_base = {}
        suffix = "." + resource["name"].rsplit(".", 1)[-1]
        for mt in DC_MIME_TYPES:
            if suffix in DC_MIME_TYPES[mt]:
                res_dict_base["mimetype"] = mt
                break

        # Also make sure the resource has "url" defined.
        site_url = get_ckan_config_option("ckan.site_url")
        meta_url = (f"{site_url}"
                    f"/dataset/{resource['package_id']}"
>                   f"/resource/{resource['id']}"
                    f"/download/{resource['name'].lower()}")
E       KeyError: 'id'

/usr/local/lib/python3.10/site-packages/ckanext/dcor_schemas/jobs.py:30: KeyError
```

RaghavaAlajangi commented:
This time, we got different errors:
```
(venv) root@dcor-dev-docker:/srv/app/src/ckanext-dc_view# coverage run --source=ckanext.dc_view --omit=*tests* -m pytest -p no:warnings ckanext
2025-04-17 12:28:36,659 INFO  [ckan.cli] Using configuration file /srv/app/ckan.ini
2025-04-17 12:28:36,659 INFO  [ckan.config.environment] Loading static files from public
2025-04-17 12:28:38,350 INFO  [ckanext.envvars.plugin] Loading new plugin: datatables_view
2025-04-17 12:28:38,559 WARNI [ckan.common] Option lang is not declared
2025-04-17 12:28:38,596 INFO  [ckanext.envvars.plugin] Loading new plugin: datastore
2025-04-17 12:28:38,847 INFO  [ckanext.envvars.plugin] Loading new plugin: datapusher
2025-04-17 12:28:39,074 INFO  [ckan.config.environment] Loading templates from /srv/app/src/ckan/ckan/templates
2025-04-17 12:28:39,236 INFO  [ckan.config.environment] Loading templates from /srv/app/src/ckan/ckan/templates
2025-04-17 12:28:39,309 INFO  [ckan.config.environment] Loading templates from /srv/app/src/ckan/ckan/templates
2025-04-17 12:28:39,398 INFO  [ckan.config.environment] Loading templates from /srv/app/src/ckan/ckan/templates
======================================================================================================== test session starts =========================================================================================================
platform linux -- Python 3.10.16, pytest-8.2.2, pluggy-1.5.0
rootdir: /srv/app/src/ckanext-dc_view
configfile: pyproject.toml
plugins: split-0.9.0, cov-5.0.0, factoryboy-2.7.0, rerunfailures-14.0, freezegun-0.4.2, Faker-26.0.0, ckan-2.11.2
collected 7 items                                                                                                                                                                                                                    
2025-04-17 12:28:39,790 INFO  [ckan.config.environment] Loading templates from /srv/app/src/ckan/ckan/templates

ckanext/dc_view/tests/test_jobs.py FF                                                                                                                                                                                          [ 28%]
ckanext/dc_view/tests/test_plugin.py .                                                                                                                                                                                         [ 42%]
ckanext/dc_view/tests/test_route.py FFFF                                                                                                                                                                                       [100%]

============================================================================================================== FAILURES ==============================================================================================================
_____________________________________________________________________________________________________ test_create_preview_s3_job _____________________________________________________________________________________________________

enqueue_job_mock = <MagicMock name='enqueue_job' id='140667794223072'>, create_with_upload = <function create_with_upload.<locals>.factory at 0x7fefc5e52cb0>
monkeypatch = <_pytest.monkeypatch.MonkeyPatch object at 0x7fefc6cf6f20>
ckan_config = {'beaker.session.validate_key': '32aa7a1d-a86c-4afe-8682-29384fc657e8', 'beaker.session.encrypt_key': 'c6017a45-9ae3-4...TRAP_HTTP_EXCEPTIONS': False, 'BABEL_DEFAULT_TIMEZONE': 'UTC', 'DEBUG_TB_HOSTS': (), 'SQLALCHEMY_RECORD_QUERIES': True}
tmpdir = local('/tmp/pytest-of-root/pytest-0/test_create_preview_s3_job0')

    @pytest.mark.ckan_config('ckan.plugins', 'dcor_depot dcor_schemas dc_view')
    @pytest.mark.usefixtures('clean_db', 'with_request_context')
    @mock.patch('ckan.plugins.toolkit.enqueue_job',
                side_effect=synchronous_enqueue_job)
    def test_create_preview_s3_job(
            enqueue_job_mock, create_with_upload, monkeypatch, ckan_config,
            tmpdir):
        monkeypatch.setitem(ckan_config, 'ckan.storage_path', str(tmpdir))
        monkeypatch.setattr(ckan.lib.uploader,
                            'get_storage_path',
                            lambda: str(tmpdir))
        monkeypatch.setattr(
            ckanext.dcor_schemas.plugin,
            'DISABLE_AFTER_DATASET_CREATE_FOR_CONCURRENT_JOB_TESTS',
            True)
    
        user = factories.User()
        owner_org = factories.Organization(users=[{
            'name': user['id'],
            'capacity': 'admin'
        }])
        # Note: `call_action` bypasses authorization!
        # create 1st dataset
        create_context = {'ignore_auth': False, 'user': user['name'],
                          'api_version': 3}
        dataset = make_dataset(create_context,
                               owner_org,
                               activate=False)
        path = data_path / "calibration_beads_47.rtdc"
        content = path.read_bytes()
>       result = create_with_upload(
            content, 'test.rtdc',
            url="upload",
            package_id=dataset["id"],
            context=create_context,
        )

ckanext/dc_view/tests/test_jobs.py:60: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../ckan/ckan/tests/pytest_ckan/fixtures.py:568: in factory
    return test_helpers.call_action(action, context, **params)
../ckan/ckan/tests/helpers.py:128: in call_action
    return logic.get_action(action_name)(context, kwargs)
../ckan/ckan/logic/__init__.py:580: in wrapped
    result = _action(context, data_dict, **kw)
../ckan/ckan/logic/action/create.py:347: in resource_create
    plugin.after_resource_create(context, resource)
/usr/local/lib/python3.10/site-packages/ckanext/dcor_depot/plugin.py:44: in after_resource_create
    jobs.RQJob.enqueue_all_jobs(resource, ckanext="dcor_depot")
/usr/local/lib/python3.10/site-packages/dcor_shared/rqjob.py:74: in enqueue_all_jobs
    inst.enqueue_job(resource, redis_connect=redis_connect)
/usr/local/lib/python3.10/site-packages/dcor_shared/rqjob.py:59: in enqueue_job
    toolkit.enqueue_job(self.method,
/usr/local/lib/python3.10/unittest/mock.py:1114: in __call__
    return self._mock_call(*args, **kwargs)
/usr/local/lib/python3.10/unittest/mock.py:1118: in _mock_call
    return self._execute_mock_call(*args, **kwargs)
/usr/local/lib/python3.10/unittest/mock.py:1179: in _execute_mock_call
    result = effect(*args, **kwargs)
/usr/local/lib/python3.10/site-packages/dcor_shared/testing.py:250: in synchronous_enqueue_job
    job_func(*args, **kwargs)
/usr/local/lib/python3.10/site-packages/ckanext/dcor_depot/jobs.py:46: in job_migrate_resource_to_s3
    wait_for_resource(rid)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

resource_id = '13cf263e-1b2e-47b6-a2c7-c8309553bf1f', timeout = 10

    @functools.lru_cache(maxsize=100)
    def wait_for_resource(resource_id: str,
                          timeout: float = 10):
        """Wait for resource to be available
    
        This function can be used by other plugins to ensure that
        a resource is available for processing.
    
        There multiple ways for data to become available:
    
        1. The ckanext-dcor_depot plugin imports data by touching
           dummy files and then sym-linking to data on disk. Here
           we just check that the file is not a dummy file anymore.
        2. Legacy uploads via nginx/uwsgi directly into CKAN and onto
           the local block storage worked the same way. We have to wait
           for the dummy file to be replaced.
        3. The new (2024) way of uploading data is via pre-signed URLs
           to an S3 instance. Here, we have to make sure that the
           upload is complete and the file exists. If this is the case,
           then uploads should have already completed when this function
           is called, so we only check for the existence of the resource
           in ckan and whether the `s3_available` attribute is defined.
        """
        from ckan.common import config
        from ckan import logic
    
        if len(resource_id) != 36:
            raise ValueError(f"Invalid resource id: {resource_id}")
    
        resource_show = logic.get_action("resource_show")
        path = pathlib.Path(get_resource_path(resource_id))
    
        dcor_depot_available = "dcor_depot" in config.get('ckan.plugins', "")
        # Initially this was set to 10s, but if `/data` is mounted on a
        # network share then this part here just takes too long.
        t0 = time.time()
        ld = len(DUMMY_BYTES)
        while True:
            try:
                res_dict = resource_show(context={'ignore_auth': True,
                                                  'user': 'default'},
                                         data_dict={"id": resource_id})
            except logic.NotFound:
                # Other processes are still working on getting the resource
                # online. We have to wait.
                time.sleep(5)
                continue
    
            s3_ok = res_dict.get("s3_available", None)
            if time.time() - t0 > timeout:
>               raise OSError("Data import seems to take too long "
E               OSError: Data import seems to take too long for '/tmp/pytest-of-root/pytest-0/test_create_preview_s3_job0/resources/13c/f26/3e-1b2e-47b6-a2c7-c8309553bf1f'!

/usr/local/lib/python3.10/site-packages/dcor_shared/data.py:72: OSError
------------------------------------------------------------------------------------------------------- Captured stderr setup --------------------------------------------------------------------------------------------------------
2025-04-17 12:28:40,161 INFO  [ckan.model] Database tables dropped
2025-04-17 12:28:41,935 INFO  [ckan.model] CKAN database version upgraded: base -> 4a5e3465beb6 (head)
2025-04-17 12:28:41,935 INFO  [ckan.model] Database initialised
2025-04-17 12:28:41,935 INFO  [ckan.model] Database rebuilt
2025-04-17 12:28:41,937 INFO  [ckan.config.environment] Loading static files from public
2025-04-17 12:28:42,052 INFO  [ckan.config.environment] Loading templates from /srv/app/src/ckan/ckan/templates
2025-04-17 12:28:42,212 INFO  [ckan.config.environment] Loading templates from /srv/app/src/ckan/ckan/templates
2025-04-17 12:28:42,238 ERROR [ckan.lib.app_globals] Asset 'dcor_theme_main/dcor_theme_main' does not exist. Fallback to 'css/main'
-------------------------------------------------------------------------------------------------------- Captured stderr call --------------------------------------------------------------------------------------------------------
2025-04-17 12:28:43,212 WARNI [ckan.lib.datapreview] Plugin for view image_view could not be found
2025-04-17 12:28:43,212 WARNI [ckan.lib.datapreview] Plugin for view text_view could not be found
2025-04-17 12:28:43,213 WARNI [ckan.lib.datapreview] Plugin for view dc_log_view could not be found
________________________________________________________________________________________________ test_create_preview_s3_job_alternate ________________________________________________________________________________________________

enqueue_job_mock = <MagicMock name='enqueue_job' id='140667789139088'>, create_with_upload = <function create_with_upload.<locals>.factory at 0x7fefc5a71fc0>
monkeypatch = <_pytest.monkeypatch.MonkeyPatch object at 0x7fefc5ae2d40>
ckan_config = {'beaker.session.validate_key': '32aa7a1d-a86c-4afe-8682-29384fc657e8', 'beaker.session.encrypt_key': 'c6017a45-9ae3-4...TRAP_HTTP_EXCEPTIONS': False, 'BABEL_DEFAULT_TIMEZONE': 'UTC', 'DEBUG_TB_HOSTS': (), 'SQLALCHEMY_RECORD_QUERIES': True}
tmpdir = local('/tmp/pytest-of-root/pytest-0/test_create_preview_s3_job_alt0')

    @pytest.mark.ckan_config('ckan.plugins', 'dcor_depot dc_view dcor_schemas')
    @pytest.mark.usefixtures('clean_db', 'with_request_context')
    @mock.patch('ckan.plugins.toolkit.enqueue_job',
                side_effect=synchronous_enqueue_job)
    def test_create_preview_s3_job_alternate(
            enqueue_job_mock, create_with_upload, monkeypatch, ckan_config,
            tmpdir):
        monkeypatch.setitem(ckan_config, 'ckan.storage_path', str(tmpdir))
        monkeypatch.setattr(ckan.lib.uploader,
                            'get_storage_path',
                            lambda: str(tmpdir))
        monkeypatch.setattr(
            ckanext.dcor_schemas.plugin,
            'DISABLE_AFTER_DATASET_CREATE_FOR_CONCURRENT_JOB_TESTS',
            True)
    
        user = factories.User()
        owner_org = factories.Organization(users=[{
            'name': user['id'],
            'capacity': 'admin'
        }])
        # Note: `call_action` bypasses authorization!
        # create 1st dataset
        create_context = {'ignore_auth': False,
                          'user': user['name'],
                          'api_version': 3}
>       ds_dict, res_dict = make_dataset(
            create_context,
            owner_org,
            create_with_upload=create_with_upload,
            resource_path=data_path / "calibration_beads_47.rtdc",
            activate=True)

ckanext/dc_view/tests/test_jobs.py:101: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/usr/local/lib/python3.10/site-packages/dcor_shared/testing.py:83: in make_dataset
    rs = make_resource(resource_path=resource_path,
/usr/local/lib/python3.10/site-packages/dcor_shared/testing.py:192: in make_resource
    rs = create_with_upload(
../ckan/ckan/tests/pytest_ckan/fixtures.py:568: in factory
    return test_helpers.call_action(action, context, **params)
../ckan/ckan/tests/helpers.py:128: in call_action
    return logic.get_action(action_name)(context, kwargs)
../ckan/ckan/logic/__init__.py:580: in wrapped
    result = _action(context, data_dict, **kw)
../ckan/ckan/logic/action/create.py:347: in resource_create
    plugin.after_resource_create(context, resource)
/usr/local/lib/python3.10/site-packages/ckanext/dcor_depot/plugin.py:44: in after_resource_create
    jobs.RQJob.enqueue_all_jobs(resource, ckanext="dcor_depot")
/usr/local/lib/python3.10/site-packages/dcor_shared/rqjob.py:74: in enqueue_all_jobs
    inst.enqueue_job(resource, redis_connect=redis_connect)
/usr/local/lib/python3.10/site-packages/dcor_shared/rqjob.py:59: in enqueue_job
    toolkit.enqueue_job(self.method,
/usr/local/lib/python3.10/unittest/mock.py:1114: in __call__
    return self._mock_call(*args, **kwargs)
/usr/local/lib/python3.10/unittest/mock.py:1118: in _mock_call
    return self._execute_mock_call(*args, **kwargs)
/usr/local/lib/python3.10/unittest/mock.py:1179: in _execute_mock_call
    result = effect(*args, **kwargs)
/usr/local/lib/python3.10/site-packages/dcor_shared/testing.py:250: in synchronous_enqueue_job
    job_func(*args, **kwargs)
/usr/local/lib/python3.10/site-packages/ckanext/dcor_depot/jobs.py:46: in job_migrate_resource_to_s3
    wait_for_resource(rid)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

resource_id = 'ad7a9522-3446-4e61-8d20-1b2df7011dbd', timeout = 10

    @functools.lru_cache(maxsize=100)
    def wait_for_resource(resource_id: str,
                          timeout: float = 10):
        """Wait for resource to be available
    
        This function can be used by other plugins to ensure that
        a resource is available for processing.
    
        There multiple ways for data to become available:
    
        1. The ckanext-dcor_depot plugin imports data by touching
           dummy files and then sym-linking to data on disk. Here
           we just check that the file is not a dummy file anymore.
        2. Legacy uploads via nginx/uwsgi directly into CKAN and onto
           the local block storage worked the same way. We have to wait
           for the dummy file to be replaced.
        3. The new (2024) way of uploading data is via pre-signed URLs
           to an S3 instance. Here, we have to make sure that the
           upload is complete and the file exists. If this is the case,
           then uploads should have already completed when this function
           is called, so we only check for the existence of the resource
           in ckan and whether the `s3_available` attribute is defined.
        """
        from ckan.common import config
        from ckan import logic
    
        if len(resource_id) != 36:
            raise ValueError(f"Invalid resource id: {resource_id}")
    
        resource_show = logic.get_action("resource_show")
        path = pathlib.Path(get_resource_path(resource_id))
    
        dcor_depot_available = "dcor_depot" in config.get('ckan.plugins', "")
        # Initially this was set to 10s, but if `/data` is mounted on a
        # network share then this part here just takes too long.
        t0 = time.time()
        ld = len(DUMMY_BYTES)
        while True:
            try:
                res_dict = resource_show(context={'ignore_auth': True,
                                                  'user': 'default'},
                                         data_dict={"id": resource_id})
            except logic.NotFound:
                # Other processes are still working on getting the resource
                # online. We have to wait.
                time.sleep(5)
                continue
    
            s3_ok = res_dict.get("s3_available", None)
            if time.time() - t0 > timeout:
>               raise OSError("Data import seems to take too long "
E               OSError: Data import seems to take too long for '/tmp/pytest-of-root/pytest-0/test_create_preview_s3_job_alt0/resources/ad7/a95/22-3446-4e61-8d20-1b2df7011dbd'!

/usr/local/lib/python3.10/site-packages/dcor_shared/data.py:72: OSError
------------------------------------------------------------------------------------------------------- Captured stderr setup --------------------------------------------------------------------------------------------------------
2025-04-17 12:28:53,557 INFO  [ckan.model] Database table data deleted
2025-04-17 12:28:53,557 INFO  [ckan.model] Database initialised
2025-04-17 12:28:53,557 INFO  [ckan.model] Database rebuilt
2025-04-17 12:28:53,558 INFO  [ckan.config.environment] Loading static files from public
2025-04-17 12:28:53,672 INFO  [ckan.config.environment] Loading templates from /srv/app/src/ckan/ckan/templates
2025-04-17 12:28:53,840 INFO  [ckan.config.environment] Loading templates from /srv/app/src/ckan/ckan/templates
2025-04-17 12:28:53,865 ERROR [ckan.lib.app_globals] Asset 'dcor_theme_main/dcor_theme_main' does not exist. Fallback to 'css/main'
-------------------------------------------------------------------------------------------------------- Captured stderr call --------------------------------------------------------------------------------------------------------
2025-04-17 12:28:54,708 WARNI [ckan.lib.datapreview] Plugin for view image_view could not be found
2025-04-17 12:28:54,709 WARNI [ckan.lib.datapreview] Plugin for view text_view could not be found
2025-04-17 12:28:54,710 WARNI [ckan.lib.datapreview] Plugin for view dc_log_view could not be found
_____________________________________________________________________________________________ test_route_redirect_preview_to_s3_private ______________________________________________________________________________________________

enqueue_job_mock = <MagicMock name='enqueue_job' id='140667799202128'>, app = <ckan.tests.helpers.CKANTestApp object at 0x7fefc59a61d0>, tmpdir = local('/tmp/pytest-of-root/pytest-0/test_route_redirect_preview_to0')
create_with_upload = <function create_with_upload.<locals>.factory at 0x7fefc5c15120>, monkeypatch = <_pytest.monkeypatch.MonkeyPatch object at 0x7fefc58e2ef0>
ckan_config = {'beaker.session.validate_key': '32aa7a1d-a86c-4afe-8682-29384fc657e8', 'beaker.session.encrypt_key': 'c6017a45-9ae3-4...TRAP_HTTP_EXCEPTIONS': False, 'BABEL_DEFAULT_TIMEZONE': 'UTC', 'DEBUG_TB_HOSTS': (), 'SQLALCHEMY_RECORD_QUERIES': True}

    @pytest.mark.ckan_config('ckan.plugins',
                             'dcor_depot dcor_schemas dc_serve dc_view')
    @pytest.mark.usefixtures('clean_db', 'with_request_context')
    @mock.patch('ckan.plugins.toolkit.enqueue_job',
                side_effect=synchronous_enqueue_job)
    def test_route_redirect_preview_to_s3_private(
            enqueue_job_mock, app, tmpdir, create_with_upload, monkeypatch,
            ckan_config):
        monkeypatch.setitem(ckan_config, 'ckan.storage_path', str(tmpdir))
        monkeypatch.setattr(ckan.lib.uploader,
                            'get_storage_path',
                            lambda: str(tmpdir))
        monkeypatch.setattr(
            ckanext.dcor_schemas.plugin,
            'DISABLE_AFTER_DATASET_CREATE_FOR_CONCURRENT_JOB_TESTS',
            True)
    
        user = factories.UserWithToken()
        user_obj = ckan.model.User.by_name(user["name"])
        monkeypatch.setattr(ckan.common,
                            'current_user',
                            user_obj)
        # Note: `call_action` bypasses authorization!
        create_context = {'ignore_auth': False,
                          'user': user['name'],
                          'api_version': 3}
        # create a dataset
>       ds_dict, res_dict = make_dataset(
            create_context=create_context,
            create_with_upload=create_with_upload,
            resource_path=data_path / "calibration_beads_47.rtdc",
            activate=True,
            private=True
            )

ckanext/dc_view/tests/test_route.py:46: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/usr/local/lib/python3.10/site-packages/dcor_shared/testing.py:83: in make_dataset
    rs = make_resource(resource_path=resource_path,
/usr/local/lib/python3.10/site-packages/dcor_shared/testing.py:192: in make_resource
    rs = create_with_upload(
../ckan/ckan/tests/pytest_ckan/fixtures.py:568: in factory
    return test_helpers.call_action(action, context, **params)
../ckan/ckan/tests/helpers.py:128: in call_action
    return logic.get_action(action_name)(context, kwargs)
../ckan/ckan/logic/__init__.py:580: in wrapped
    result = _action(context, data_dict, **kw)
../ckan/ckan/logic/action/create.py:347: in resource_create
    plugin.after_resource_create(context, resource)
/usr/local/lib/python3.10/site-packages/ckanext/dcor_depot/plugin.py:44: in after_resource_create
    jobs.RQJob.enqueue_all_jobs(resource, ckanext="dcor_depot")
/usr/local/lib/python3.10/site-packages/dcor_shared/rqjob.py:74: in enqueue_all_jobs
    inst.enqueue_job(resource, redis_connect=redis_connect)
/usr/local/lib/python3.10/site-packages/dcor_shared/rqjob.py:59: in enqueue_job
    toolkit.enqueue_job(self.method,
/usr/local/lib/python3.10/unittest/mock.py:1114: in __call__
    return self._mock_call(*args, **kwargs)
/usr/local/lib/python3.10/unittest/mock.py:1118: in _mock_call
    return self._execute_mock_call(*args, **kwargs)
/usr/local/lib/python3.10/unittest/mock.py:1179: in _execute_mock_call
    result = effect(*args, **kwargs)
/usr/local/lib/python3.10/site-packages/dcor_shared/testing.py:250: in synchronous_enqueue_job
    job_func(*args, **kwargs)
/usr/local/lib/python3.10/site-packages/ckanext/dcor_depot/jobs.py:46: in job_migrate_resource_to_s3
    wait_for_resource(rid)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

resource_id = '498f82cf-ff86-4c55-bbe5-1c9b1efd2fbb', timeout = 10

    @functools.lru_cache(maxsize=100)
    def wait_for_resource(resource_id: str,
                          timeout: float = 10):
        """Wait for resource to be available
    
        This function can be used by other plugins to ensure that
        a resource is available for processing.
    
        There multiple ways for data to become available:
    
        1. The ckanext-dcor_depot plugin imports data by touching
           dummy files and then sym-linking to data on disk. Here
           we just check that the file is not a dummy file anymore.
        2. Legacy uploads via nginx/uwsgi directly into CKAN and onto
           the local block storage worked the same way. We have to wait
           for the dummy file to be replaced.
        3. The new (2024) way of uploading data is via pre-signed URLs
           to an S3 instance. Here, we have to make sure that the
           upload is complete and the file exists. If this is the case,
           then uploads should have already completed when this function
           is called, so we only check for the existence of the resource
           in ckan and whether the `s3_available` attribute is defined.
        """
        from ckan.common import config
        from ckan import logic
    
        if len(resource_id) != 36:
            raise ValueError(f"Invalid resource id: {resource_id}")
    
        resource_show = logic.get_action("resource_show")
        path = pathlib.Path(get_resource_path(resource_id))
    
        dcor_depot_available = "dcor_depot" in config.get('ckan.plugins', "")
        # Initially this was set to 10s, but if `/data` is mounted on a
        # network share then this part here just takes too long.
        t0 = time.time()
        ld = len(DUMMY_BYTES)
        while True:
            try:
                res_dict = resource_show(context={'ignore_auth': True,
                                                  'user': 'default'},
                                         data_dict={"id": resource_id})
            except logic.NotFound:
                # Other processes are still working on getting the resource
                # online. We have to wait.
                time.sleep(5)
                continue
    
            s3_ok = res_dict.get("s3_available", None)
            if time.time() - t0 > timeout:
>               raise OSError("Data import seems to take too long "
E               OSError: Data import seems to take too long for '/tmp/pytest-of-root/pytest-0/test_route_redirect_preview_to0/resources/498/f82/cf-ff86-4c55-bbe5-1c9b1efd2fbb'!

/usr/local/lib/python3.10/site-packages/dcor_shared/data.py:72: OSError
------------------------------------------------------------------------------------------------------- Captured stderr setup --------------------------------------------------------------------------------------------------------
2025-04-17 12:29:05,057 INFO  [ckan.model] Database table data deleted
2025-04-17 12:29:05,057 INFO  [ckan.model] Database initialised
2025-04-17 12:29:05,057 INFO  [ckan.model] Database rebuilt
2025-04-17 12:29:05,059 INFO  [ckan.config.environment] Loading static files from public
2025-04-17 12:29:05,207 INFO  [ckan.config.environment] Loading templates from /srv/app/src/ckan/ckan/templates
2025-04-17 12:29:05,381 INFO  [ckan.config.environment] Loading templates from /srv/app/src/ckan/ckan/templates
2025-04-17 12:29:05,425 ERROR [ckan.lib.app_globals] Asset 'dcor_theme_main/dcor_theme_main' does not exist. Fallback to 'css/main'
-------------------------------------------------------------------------------------------------------- Captured stderr call --------------------------------------------------------------------------------------------------------
2025-04-17 12:29:06,352 WARNI [ckan.lib.datapreview] Plugin for view image_view could not be found
2025-04-17 12:29:06,352 WARNI [ckan.lib.datapreview] Plugin for view text_view could not be found
2025-04-17 12:29:06,353 WARNI [ckan.lib.datapreview] Plugin for view dc_log_view could not be found
__________________________________________________________________________________________________ test_route_preview_to_s3_public ___________________________________________________________________________________________________

enqueue_job_mock = <MagicMock name='enqueue_job' id='140667792775808'>, app = <ckan.tests.helpers.CKANTestApp object at 0x7fefc6343dc0>, tmpdir = local('/tmp/pytest-of-root/pytest-0/test_route_preview_to_s3_publi0')
create_with_upload = <function create_with_upload.<locals>.factory at 0x7fefc59888b0>, monkeypatch = <_pytest.monkeypatch.MonkeyPatch object at 0x7fefc55ffe50>
ckan_config = {'beaker.session.validate_key': '32aa7a1d-a86c-4afe-8682-29384fc657e8', 'beaker.session.encrypt_key': 'c6017a45-9ae3-4...TRAP_HTTP_EXCEPTIONS': False, 'BABEL_DEFAULT_TIMEZONE': 'UTC', 'DEBUG_TB_HOSTS': (), 'SQLALCHEMY_RECORD_QUERIES': True}

    @pytest.mark.ckan_config('ckan.plugins',
                             'dcor_depot dcor_schemas dc_serve dc_view')
    @pytest.mark.usefixtures('clean_db', 'with_request_context')
    @mock.patch('ckan.plugins.toolkit.enqueue_job',
                side_effect=synchronous_enqueue_job)
    def test_route_preview_to_s3_public(
            enqueue_job_mock, app, tmpdir, create_with_upload, monkeypatch,
            ckan_config):
        monkeypatch.setitem(ckan_config, 'ckan.storage_path', str(tmpdir))
        monkeypatch.setattr(ckan.lib.uploader,
                            'get_storage_path',
                            lambda: str(tmpdir))
        monkeypatch.setattr(
            ckanext.dcor_schemas.plugin,
            'DISABLE_AFTER_DATASET_CREATE_FOR_CONCURRENT_JOB_TESTS',
            True)
    
        # create a dataset
>       ds_dict, res_dict = make_dataset(
            create_with_upload=create_with_upload,
            resource_path=data_path / "calibration_beads_47.rtdc",
            activate=True)

ckanext/dc_view/tests/test_route.py:108: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/usr/local/lib/python3.10/site-packages/dcor_shared/testing.py:83: in make_dataset
    rs = make_resource(resource_path=resource_path,
/usr/local/lib/python3.10/site-packages/dcor_shared/testing.py:192: in make_resource
    rs = create_with_upload(
../ckan/ckan/tests/pytest_ckan/fixtures.py:568: in factory
    return test_helpers.call_action(action, context, **params)
../ckan/ckan/tests/helpers.py:128: in call_action
    return logic.get_action(action_name)(context, kwargs)
../ckan/ckan/logic/__init__.py:580: in wrapped
    result = _action(context, data_dict, **kw)
../ckan/ckan/logic/action/create.py:347: in resource_create
    plugin.after_resource_create(context, resource)
/usr/local/lib/python3.10/site-packages/ckanext/dcor_depot/plugin.py:44: in after_resource_create
    jobs.RQJob.enqueue_all_jobs(resource, ckanext="dcor_depot")
/usr/local/lib/python3.10/site-packages/dcor_shared/rqjob.py:74: in enqueue_all_jobs
    inst.enqueue_job(resource, redis_connect=redis_connect)
/usr/local/lib/python3.10/site-packages/dcor_shared/rqjob.py:59: in enqueue_job
    toolkit.enqueue_job(self.method,
/usr/local/lib/python3.10/unittest/mock.py:1114: in __call__
    return self._mock_call(*args, **kwargs)
/usr/local/lib/python3.10/unittest/mock.py:1118: in _mock_call
    return self._execute_mock_call(*args, **kwargs)
/usr/local/lib/python3.10/unittest/mock.py:1179: in _execute_mock_call
    result = effect(*args, **kwargs)
/usr/local/lib/python3.10/site-packages/dcor_shared/testing.py:250: in synchronous_enqueue_job
    job_func(*args, **kwargs)
/usr/local/lib/python3.10/site-packages/ckanext/dcor_depot/jobs.py:46: in job_migrate_resource_to_s3
    wait_for_resource(rid)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

resource_id = '952a7e32-9336-4558-80d5-2ed1c2a71d0d', timeout = 10

    @functools.lru_cache(maxsize=100)
    def wait_for_resource(resource_id: str,
                          timeout: float = 10):
        """Wait for resource to be available
    
        This function can be used by other plugins to ensure that
        a resource is available for processing.
    
        There multiple ways for data to become available:
    
        1. The ckanext-dcor_depot plugin imports data by touching
           dummy files and then sym-linking to data on disk. Here
           we just check that the file is not a dummy file anymore.
        2. Legacy uploads via nginx/uwsgi directly into CKAN and onto
           the local block storage worked the same way. We have to wait
           for the dummy file to be replaced.
        3. The new (2024) way of uploading data is via pre-signed URLs
           to an S3 instance. Here, we have to make sure that the
           upload is complete and the file exists. If this is the case,
           then uploads should have already completed when this function
           is called, so we only check for the existence of the resource
           in ckan and whether the `s3_available` attribute is defined.
        """
        from ckan.common import config
        from ckan import logic
    
        if len(resource_id) != 36:
            raise ValueError(f"Invalid resource id: {resource_id}")
    
        resource_show = logic.get_action("resource_show")
        path = pathlib.Path(get_resource_path(resource_id))
    
        dcor_depot_available = "dcor_depot" in config.get('ckan.plugins', "")
        # Initially this was set to 10s, but if `/data` is mounted on a
        # network share then this part here just takes too long.
        t0 = time.time()
        ld = len(DUMMY_BYTES)
        while True:
            try:
                res_dict = resource_show(context={'ignore_auth': True,
                                                  'user': 'default'},
                                         data_dict={"id": resource_id})
            except logic.NotFound:
                # Other processes are still working on getting the resource
                # online. We have to wait.
                time.sleep(5)
                continue
    
            s3_ok = res_dict.get("s3_available", None)
            if time.time() - t0 > timeout:
>               raise OSError("Data import seems to take too long "
E               OSError: Data import seems to take too long for '/tmp/pytest-of-root/pytest-0/test_route_preview_to_s3_publi0/resources/952/a7e/32-9336-4558-80d5-2ed1c2a71d0d'!

/usr/local/lib/python3.10/site-packages/dcor_shared/data.py:72: OSError
------------------------------------------------------------------------------------------------------- Captured stderr setup --------------------------------------------------------------------------------------------------------
2025-04-17 12:29:16,586 INFO  [ckan.model] Database table data deleted
2025-04-17 12:29:16,586 INFO  [ckan.model] Database initialised
2025-04-17 12:29:16,586 INFO  [ckan.model] Database rebuilt
2025-04-17 12:29:16,588 INFO  [ckan.config.environment] Loading static files from public
2025-04-17 12:29:16,701 INFO  [ckan.config.environment] Loading templates from /srv/app/src/ckan/ckan/templates
2025-04-17 12:29:16,953 INFO  [ckan.config.environment] Loading templates from /srv/app/src/ckan/ckan/templates
2025-04-17 12:29:16,983 ERROR [ckan.lib.app_globals] Asset 'dcor_theme_main/dcor_theme_main' does not exist. Fallback to 'css/main'
-------------------------------------------------------------------------------------------------------- Captured stderr call --------------------------------------------------------------------------------------------------------
2025-04-17 12:29:17,876 WARNI [ckan.lib.datapreview] Plugin for view image_view could not be found
2025-04-17 12:29:17,877 WARNI [ckan.lib.datapreview] Plugin for view text_view could not be found
2025-04-17 12:29:17,878 WARNI [ckan.lib.datapreview] Plugin for view dc_log_view could not be found
____________________________________________________________________________________________ test_route_s3_redirect_preview_to_s3_private ____________________________________________________________________________________________

enqueue_job_mock = <MagicMock name='enqueue_job' id='140667799209440'>, app = <ckan.tests.helpers.CKANTestApp object at 0x7fefc5d1d060>, tmpdir = local('/tmp/pytest-of-root/pytest-0/test_route_s3_redirect_preview0')
monkeypatch = <_pytest.monkeypatch.MonkeyPatch object at 0x7fefc586da80>
ckan_config = {'beaker.session.validate_key': '32aa7a1d-a86c-4afe-8682-29384fc657e8', 'beaker.session.encrypt_key': 'c6017a45-9ae3-4...TRAP_HTTP_EXCEPTIONS': False, 'BABEL_DEFAULT_TIMEZONE': 'UTC', 'DEBUG_TB_HOSTS': (), 'SQLALCHEMY_RECORD_QUERIES': True}

    @pytest.mark.ckan_config('ckan.plugins',
                             'dcor_depot dcor_schemas dc_serve dc_view')
    @pytest.mark.usefixtures('clean_db', 'with_request_context')
    @mock.patch('ckan.plugins.toolkit.enqueue_job',
                side_effect=synchronous_enqueue_job)
    def test_route_s3_redirect_preview_to_s3_private(
            enqueue_job_mock, app, tmpdir, monkeypatch,
            ckan_config):
        monkeypatch.setitem(ckan_config, 'ckan.storage_path', str(tmpdir))
        monkeypatch.setattr(ckan.lib.uploader,
                            'get_storage_path',
                            lambda: str(tmpdir))
    
        user = factories.UserWithToken()
        user_obj = ckan.model.User.by_name(user["name"])
        monkeypatch.setattr(ckan.common,
                            'current_user',
                            user_obj)
    
        # Note: `call_action` bypasses authorization!
        create_context = {'ignore_auth': False,
                          'user': user['name'],
                          'api_version': 3}
        # create a dataset
        ds_dict, res_dict = make_dataset_via_s3(
            create_context=create_context,
            resource_path=data_path / "calibration_beads_47.rtdc",
            activate=True,
            private=True
            )
        rid = res_dict["id"]
        assert "s3_available" in res_dict
        assert "s3_url" in res_dict
        assert len(res_dict.get("url"))
    
        # Since version 0.9.0, we are no longer storing the preview image locally.
        path = dcor_shared.get_resource_path(rid)
        path_prev = path.with_name(path.name + "_preview.jpg")
        assert not path_prev.exists(), "sanity check"
    
        did = ds_dict["id"]
        # We should not be authorized to access the resource without API token
        resp0 = app.get(
            f"/dataset/{did}/resource/{rid}/preview.jpg",
            status=404
            )
        assert len(resp0.history) == 0
    
        # Try again with token
        resp = app.get(
            f"/dataset/{did}/resource/{rid}/preview.jpg",
            headers={u"authorization": user["token"]},
            )
    
        endpoint = dcor_shared.get_ckan_config_option(
            "dcor_object_store.endpoint_url")
        bucket_name = dcor_shared.get_ckan_config_option(
            "dcor_object_store.bucket_name").format(
            organization_id=ds_dict["organization"]["id"])
>       redirect = resp.history[0]
E       IndexError: tuple index out of range

ckanext/dc_view/tests/test_route.py:196: IndexError
------------------------------------------------------------------------------------------------------- Captured stderr setup --------------------------------------------------------------------------------------------------------
2025-04-17 12:29:28,127 INFO  [ckan.model] Database table data deleted
2025-04-17 12:29:28,127 INFO  [ckan.model] Database initialised
2025-04-17 12:29:28,127 INFO  [ckan.model] Database rebuilt
2025-04-17 12:29:28,129 INFO  [ckan.config.environment] Loading static files from public
2025-04-17 12:29:28,253 INFO  [ckan.config.environment] Loading templates from /srv/app/src/ckan/ckan/templates
2025-04-17 12:29:28,439 INFO  [ckan.config.environment] Loading templates from /srv/app/src/ckan/ckan/templates
2025-04-17 12:29:28,467 ERROR [ckan.lib.app_globals] Asset 'dcor_theme_main/dcor_theme_main' does not exist. Fallback to 'css/main'
-------------------------------------------------------------------------------------------------------- Captured stdout call --------------------------------------------------------------------------------------------------------
Using ancillary features ['volume']
-------------------------------------------------------------------------------------------------------- Captured stderr call --------------------------------------------------------------------------------------------------------
2025-04-17 12:29:31,063 WARNI [ckan.lib.datapreview] Plugin for view image_view could not be found
2025-04-17 12:29:31,063 WARNI [ckan.lib.datapreview] Plugin for view text_view could not be found
2025-04-17 12:29:31,064 WARNI [ckan.lib.datapreview] Plugin for view dc_log_view could not be found
2025-04-17 12:29:31,074 INFO  [ckanext.dc_serve.jobs] Generating condensed resource fc70ab58-9cd4-483e-be7b-422f0252f024
2025-04-17 12:29:31,078 WARNI [ckan.common] Option ckanext.dc_serve.tmp_dir is not declared
2025-04-17 12:29:33,130 INFO  [ckan.config.middleware.flask_app]  404 /dataset/cf0b15a0-1433-4c84-9a15-c653fd003ef0/resource/fc70ab58-9cd4-483e-be7b-422f0252f024/preview.jpg render time 0.515 seconds
2025-04-17 12:29:33,288 INFO  [ckan.config.middleware.flask_app]  404 /dataset/cf0b15a0-1433-4c84-9a15-c653fd003ef0/resource/fc70ab58-9cd4-483e-be7b-422f0252f024/preview.jpg render time 0.016 seconds
_________________________________________________________________________________________________ test_route_s3_preview_to_s3_public _________________________________________________________________________________________________

enqueue_job_mock = <MagicMock name='enqueue_job' id='140667514640400'>, app = <ckan.tests.helpers.CKANTestApp object at 0x7fefc6343ee0>, tmpdir = local('/tmp/pytest-of-root/pytest-0/test_route_s3_preview_to_s3_pu0')
monkeypatch = <_pytest.monkeypatch.MonkeyPatch object at 0x7fefbf394c70>
ckan_config = {'beaker.session.validate_key': '32aa7a1d-a86c-4afe-8682-29384fc657e8', 'beaker.session.encrypt_key': 'c6017a45-9ae3-4...TRAP_HTTP_EXCEPTIONS': False, 'BABEL_DEFAULT_TIMEZONE': 'UTC', 'DEBUG_TB_HOSTS': (), 'SQLALCHEMY_RECORD_QUERIES': True}

    @pytest.mark.ckan_config('ckan.plugins',
                             'dcor_depot dcor_schemas dc_serve dc_view')
    @pytest.mark.usefixtures('clean_db', 'with_request_context')
    @mock.patch('ckan.plugins.toolkit.enqueue_job',
                side_effect=synchronous_enqueue_job)
    def test_route_s3_preview_to_s3_public(
            enqueue_job_mock, app, tmpdir, monkeypatch,
            ckan_config):
        monkeypatch.setitem(ckan_config, 'ckan.storage_path', str(tmpdir))
        monkeypatch.setattr(ckan.lib.uploader,
                            'get_storage_path',
                            lambda: str(tmpdir))
    
        # create a dataset
        ds_dict, res_dict = make_dataset_via_s3(
            resource_path=data_path / "calibration_beads_47.rtdc",
            activate=True)
        rid = res_dict["id"]
        assert "s3_available" in res_dict
        assert "s3_url" in res_dict
        assert len(res_dict.get("url"))
    
        # Since version 0.9.0, we are no longer storing the preview image locally.
        path = dcor_shared.get_resource_path(rid)
        path_prev = path.with_name(path.name + "_preview.jpg")
        assert not path_prev.exists(), "sanity check"
    
        did = ds_dict["id"]
>       resp = app.get(
            f"/dataset/{did}/resource/{rid}/preview.jpg",
            )

ckanext/dc_view/tests/test_route.py:233: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../ckan/ckan/tests/helpers.py:222: in get
    res = self.test_client().get(url, *args, **kwargs)
/usr/local/lib/python3.10/site-packages/werkzeug/test.py:1162: in get
    return self.open(*args, **kw)
../ckan/ckan/tests/helpers.py:248: in open
    res = super(CKANTestClient, self).open(*args, **kwargs)
/usr/local/lib/python3.10/site-packages/werkzeug/test.py:1150: in open
    response = self.resolve_redirect(response, buffered=buffered)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <CKANTestClient <ckan.config.middleware.flask_app.I18nMiddleware object at 0x7fefb5392680>>, response = <WrapperTestResponse 799 bytes [302 FOUND]>, buffered = False

    def resolve_redirect(
        self, response: TestResponse, buffered: bool = False
    ) -> TestResponse:
        """Perform a new request to the location given by the redirect
        response to the previous request.
    
        :meta private:
        """
        scheme, netloc, path, qs, anchor = urlsplit(response.location)
        builder = EnvironBuilder.from_environ(
            response.request.environ, path=path, query_string=qs
        )
    
        to_name_parts = netloc.split(":", 1)[0].split(".")
        from_name_parts = builder.server_name.split(".")
    
        if to_name_parts != [""]:
            # The new location has a host, use it for the base URL.
            builder.url_scheme = scheme
            builder.host = netloc
        else:
            # A local redirect with autocorrect_location_header=False
            # doesn't have a host, so use the request's host.
            to_name_parts = from_name_parts
    
        # Explain why a redirect to a different server name won't be followed.
        if to_name_parts != from_name_parts:
            if to_name_parts[-len(from_name_parts) :] == from_name_parts:
                if not self.allow_subdomain_redirects:
                    raise RuntimeError("Following subdomain redirects is not enabled.")
            else:
>               raise RuntimeError("Following external redirects is not supported.")
E               RuntimeError: Following external redirects is not supported.

/usr/local/lib/python3.10/site-packages/werkzeug/test.py:1026: RuntimeError
------------------------------------------------------------------------------------------------------- Captured stderr setup --------------------------------------------------------------------------------------------------------
2025-04-17 12:29:33,426 INFO  [ckan.model] Database table data deleted
2025-04-17 12:29:33,427 INFO  [ckan.model] Database initialised
2025-04-17 12:29:33,427 INFO  [ckan.model] Database rebuilt
2025-04-17 12:29:33,428 INFO  [ckan.config.environment] Loading static files from public
2025-04-17 12:29:33,553 INFO  [ckan.config.environment] Loading templates from /srv/app/src/ckan/ckan/templates
2025-04-17 12:29:33,709 INFO  [ckan.config.environment] Loading templates from /srv/app/src/ckan/ckan/templates
2025-04-17 12:29:33,735 ERROR [ckan.lib.app_globals] Asset 'dcor_theme_main/dcor_theme_main' does not exist. Fallback to 'css/main'
-------------------------------------------------------------------------------------------------------- Captured stdout call --------------------------------------------------------------------------------------------------------
Using ancillary features ['volume']
-------------------------------------------------------------------------------------------------------- Captured stderr call --------------------------------------------------------------------------------------------------------
2025-04-17 12:29:36,229 WARNI [ckan.lib.datapreview] Plugin for view image_view could not be found
2025-04-17 12:29:36,229 WARNI [ckan.lib.datapreview] Plugin for view text_view could not be found
2025-04-17 12:29:36,230 WARNI [ckan.lib.datapreview] Plugin for view dc_log_view could not be found
2025-04-17 12:29:36,241 INFO  [ckanext.dc_serve.jobs] Generating condensed resource ef566ff4-0592-4e8c-b422-da981461d7ec
2025-04-17 12:29:36,245 WARNI [ckan.common] Option ckanext.dc_serve.tmp_dir is not declared
2025-04-17 12:29:37,847 INFO  [ckan.config.middleware.flask_app]  302 /dataset/e98790f4-f189-4538-9f24-0310454176f4/resource/ef566ff4-0592-4e8c-b422-da981461d7ec/preview.jpg render time 0.024 seconds
====================================================================================================== short test summary info =======================================================================================================
FAILED ckanext/dc_view/tests/test_jobs.py::test_create_preview_s3_job - OSError: Data import seems to take too long for '/tmp/pytest-of-root/pytest-0/test_create_preview_s3_job0/resources/13c/f26/3e-1b2e-47b6-a2c7-c8309553bf1f'!
FAILED ckanext/dc_view/tests/test_jobs.py::test_create_preview_s3_job_alternate - OSError: Data import seems to take too long for '/tmp/pytest-of-root/pytest-0/test_create_preview_s3_job_alt0/resources/ad7/a95/22-3446-4e61-8d20-1b2df7011dbd'!
FAILED ckanext/dc_view/tests/test_route.py::test_route_redirect_preview_to_s3_private - OSError: Data import seems to take too long for '/tmp/pytest-of-root/pytest-0/test_route_redirect_preview_to0/resources/498/f82/cf-ff86-4c55-bbe5-1c9b1efd2fbb'!
FAILED ckanext/dc_view/tests/test_route.py::test_route_preview_to_s3_public - OSError: Data import seems to take too long for '/tmp/pytest-of-root/pytest-0/test_route_preview_to_s3_publi0/resources/952/a7e/32-9336-4558-80d5-2ed1c2a71d0d'!
FAILED ckanext/dc_view/tests/test_route.py::test_route_s3_redirect_preview_to_s3_private - IndexError: tuple index out of range
FAILED ckanext/dc_view/tests/test_route.py::test_route_s3_preview_to_s3_public - RuntimeError: Following external redirects is not supported.
```

RaghavaAlajangi commented:
I also ran tests locally inside the container using `venv` and directly in the CKAN environment. After running the tests directly in the CKAN environment, the container breaks. But the `venv` approach is stable.

paulmueller commented:
I don't understand the `RuntimeError: Following external redirects is not supported.` error. I tried to fix it with 7f316bd0dd928695ce24c4fb313480cab94074c4 but it looks like there is something fundamental that prevents the testing app to follow 302 redirects.

The `OSError: Data import seems to take too long` error hints at missing background jobs. Normally, DCOR relies on background jobs that are run via supervisord. I don't know how this is solved in the docker setup. But I believe if we get background jobs running, then this will be solved. Here is the relevant documentation: https://docs.ckan.org/en/2.11/maintaining/background-tasks.html#running-background-jobs . Are background jobs part of the CKAN docker image @RaghavaAlajangi ?

RaghavaAlajangi commented:
https://github.com/ckan/ckan-docker-base/issues/66 and https://github.com/ckan/ckan-docker/issues/118
In these two issues they talked about background jobs. They mentioned that they have worker in the same ckan container and wanted to have seperate docker container for worker.

I saw this command in the issue to activate the worker. 
`ckan -c /etc/ckan/default/ckan.ini jobs worker`

Do we have to run this command manually? 
Do you execute something like this through `dcor_control`?

paulmueller commented:
These two issues are unrelated.
I could not find any evidence of background jobs being run. `supervisorctl` does not seem to be a valid command in the CKAN docker container, so it probably is not there.

In any case, the tests are running with the `enqueue_mock_job` workaround, so that background jobs don't actually have to run in the background. As a result, we don't rely on background jobs at all.

paulmueller commented:
I managed to refactor the tests so that they pass now.

@RaghavaAlajangi Please go through the docker-related changes in this PR and make sure everything is in order. After you give your OK, I will merge this and we can move to the next repository :+1:

RaghavaAlajangi commented:
Looks good to me. You can merge it.
